quilt/tools/command.py

```diff
@@ -51,25 +52,80 @@
     pass
 
 
-DEFAULT_QUILT_PKG_URL = 'https://pkg.quiltdata.com'
-QUILT_PKG_URL = os.environ.get('QUILT_PKG_URL', DEFAULT_QUILT_PKG_URL)
+DEFAULT_REGISTRY_URL = 'https://pkg.quiltdata.com'
 GIT_URL_RE = re.compile(r'(?P<url>http[s]?://[\w./~_-]+\.git)(?:@(?P<branch>[\w_-]+))?')
 
-if QUILT_PKG_URL == DEFAULT_QUILT_PKG_URL:
-    AUTH_FILE_NAME = "auth.json"
-else:
-    # Store different servers' auth in different files.
-    import hashlib
-    AUTH_FILE_NAME = "auth-%.8s.json" % hashlib.md5(QUILT_PKG_URL.encode('utf-8')).hexdigest()
-
 CHUNK_SIZE = 4096
 
 PARALLEL_UPLOADS = 20
 
 VERSION = pkg_resources.require('quilt')[0].version
 
+
+_registry_url = None
+
+def _load_config():
+    config_path = os.path.join(CONFIG_DIR, 'config.json')
+    if os.path.exists(config_path):
+        with open(config_path) as fd:
+            return json.load(fd)
+    return {}
+
+def _save_config(cfg):
+    if not os.path.exists(CONFIG_DIR):
+        os.makedirs(CONFIG_DIR)
+    config_path = os.path.join(CONFIG_DIR, 'config.json')
+    with open(config_path, 'w') as fd:
+        json.dump(cfg, fd)
+
+def get_registry_url():
+    global _registry_url
+    if _registry_url is not None:
+        return _registry_url
+
+    # Env variable; overrides the config.
+    url = os.environ.get('QUILT_PKG_URL')
+    if url is None:
+        # Config file (generated by `quilt config`).
+        cfg = _load_config()
+        url = cfg.get('registry_url', '')
+
+    # '' means default URL.
+    _registry_url = url or DEFAULT_REGISTRY_URL
+    return _registry_url
+
+def config():
+    answer = input("Please enter the URL for your custom Quilt registry (ask your administrator),\n" +
+                   "or leave this line blank to use the default registry: ")
+    if answer:
+        url = urlparse(answer.rstrip('/'))
+        if (url.scheme not in ['http', 'https'] or not url.netloc or
+            url.path or url.params or url.query or url.fragment):
+            raise CommandException("Invalid URL: %s" % answer)
+        canonical_url = urlunparse(url)
+    else:
+        canonical_url = ''  # When saving the config, store '' instead of the actual URL in case we ever change it.
+
+    cfg = _load_config()
+    cfg['registry_url'] = canonical_url
+    _save_config(cfg)
+
+    # Clear the cached URL.
+    global _registry_url
+    _registry_url = None
+
+def get_auth_path():
+    url = get_registry_url()
+    if url == DEFAULT_REGISTRY_URL:
+        suffix = ''
+    else:
+        # Store different servers' auth in different files.
+        suffix = "-%.8s" % hashlib.md5(url.encode('utf-8')).hexdigest()
+
+    return os.path.join(BASE_DIR, 'auth%s.json' % suffix)
```

Comment:
Shouldn't the config and the login/auth be stored in the same folder? Both seem like user_config rather than user_data.

Reply:
Well, it's kinda subjective. I was thinking of auth as "state" (that the user should never touch directly), while the registry URL (and soon analytics switch) is "config" that the user gets to edit.

That said, two different directories might be overkill. If we use one, it's probably better to keep using `BASE_DIR` - it's `user_data_dir`, so somewhat neutral, I guess.

Reply:
Let's keep it simple and use a single directory. That'll make it easy for someone, including us, to wipe out our config and auth data and start fresh.